### DISCUSSION TITLE
OpenSSL license with static libarary

### DISCUSSION THREAD

**Discussion by ShackGH** [openssl/openssl#24422] [Q&A] [answered]
I'm creating a static C library which includes SHA256 encryption from the OpenSSL library. As I couldn't find a way to link my static lib with the `libcrypto.a` (which has the code that I'm using), I unzipped the whole libcrypto.a lib and created a static lib containing my lib and the object files of `libcrypto.a`.
Does anyone know if this will violate any copyright laws?

**Comment by nhorman** [comment]
I can't speak to the copyright issue, but I don't think extracting and re-encoding the archive is the right technical approach here.  I think the right thing to do, if your library has a dependency on libcrypto.a is to document that, and require that linking applications link against both your library and libcrypto

**Reply by ShackGH** [reply]
It's a bit complicated as the static C lib is linked against a NativeAOT C# project

**Comment by mattcaswell** [comment, marked as answer]
I am not a lawyer and I cannot get you legal advice - you need to consult an actual lawyer for that. AFAIK there are no lawyers likely to look at these discussions - just techies :-)

However, I can point you at important parts of the licence and suggestions about what you need to consider.

Firstly, what version of OpenSSL are you using? OpenSSL prior to 3.0 comes with different licence terms to 3.0 and above. I will assume you are using 3.0 or above since those are the only versions of OpenSSL currently supported by the project.

For 3.0 and above, OpenSSL uses the Apache v2 licence. You can read it here (it also comes bundled with the OpenSSL download):

https://www.openssl.org/source/apache-license-2.0.txt

The Apache v2 licence grants you a copyright licence to do certain things:

>    2. Grant of Copyright License. Subject to the terms and conditions of
>       this License, each Contributor hereby grants to You a perpetual,
>       worldwide, non-exclusive, no-charge, royalty-free, irrevocable
>       copyright license to reproduce, prepare Derivative Works of,
>       publicly display, publicly perform, sublicense, and distribute the
>       Work and such Derivative Works in Source or Object form.

From my limited understanding of what you are doing it sounds like you are preparing a derivative work in object form.

You also have certain obligations if you are intending to redistribute your derivative work:

>    4. Redistribution. You may reproduce and distribute copies of the
>       Work or Derivative Works thereof in any medium, with or without
>       modifications, and in Source or Object form, provided that You
>       meet the following conditions:
> 
>       (a) You must give any other recipients of the Work or
>           Derivative Works a copy of this License; and
> 
>       (b) You must cause any modified files to carry prominent notices
>           stating that You changed the files; and
> 
>       (c) You must retain, in the Source form of any Derivative Works
>           that You distribute, all copyright, patent, trademark, and
>           attribution notices from the Source form of the Work,
>           excluding those notices that do not pertain to any part of
>           the Derivative Works; and
> 
>       (d) If the Work includes a "NOTICE" text file as part of its
>           distribution, then any Derivative Works that You distribute must
>           include a readable copy of the attribution notices contained
>           within such NOTICE file, excluding those notices that do not
>           pertain to any part of the Derivative Works, in at least one
>           of the following places: within a NOTICE text file distributed
>           as part of the Derivative Works; within the Source form or
>           documentation, if provided along with the Derivative Works; or,
>           within a display generated by the Derivative Works, if and
>           wherever such third-party notices normally appear. The contents
>           of the NOTICE file are for informational purposes only and
>           do not modify the License. You may add Your own attribution
>           notices within Derivative Works that You distribute, alongside
>           or as an addendum to the NOTICE text from the Work, provided
>           that such additional attribution notices cannot be construed
>           as modifying the License.
> 
>       You may add Your own copyright statement to Your modifications and
>       may provide additional or different license terms and conditions
>       for use, reproduction, or distribution of Your modifications, or
>       for any such Derivative Works as a whole, provided Your use,
>       reproduction, and distribution of the Work otherwise complies with
>       the conditions stated in this License.

So you must adhere to all of these items in your redistributed work. Most of this is non-technical in nature and doesn't say anything about the nature or form of any derivative work.

Given all of the above you need to decide for yourself whether what you are doing complies with the licence. We cannot give you that advice.

Of course I've only highlighted some specific sections that seem particularly relevant to your question. It goes without saying that you must adhere to the entire licence text.
